toc/rfc/rfc-draft-generic-per-route-features.md

```diff
@@ -0,0 +1,127 @@
+# Meta
+[meta]: #meta
+- Name: Generic Per-Route Features
+- Start Date: 2024-01-08
+- Author(s): @maxmoehl, @domdom82
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (fill in with PR link after you submit it)
+
+## Summary
+
+## Problem
+
+Implementing per-route features in Cloud Foundry is a tedious process because it requires
+changes to multiple components in multiple working groups. Most components will only need to be
+made aware of new fields without adding any significant logic but the overhead of making the
+change still exists.
+
+On the other side are the components that actually need to be aware of the change. Mostly the
+routing components (gorouter, maybe route-emitter / -registrar), the API (cloud controller)
+and UIs (CLI).
+
+## Proposal
+
+To reduce the overall overhead of implementing new per-route features we propose to add a generic
+key-value map to the definition of a route across all components. All components handling the map
+SHOULD ignore the contents of the map except gorouter, cloud controller and cf CLI which MUST handle
+the map and its values accordingly. Other components MAY limit the size of the map or size of
+keys / values for technical reasons.
+
+The following constraints apply (types are as specified in [RFC 8259](https://rfc-editor.org/rfc/rfc8259)):
+* The map MUST be representable as an object.
+* The map MUST only use strings as keys.
+* The map MUST only use numbers, strings and the literal values `true` and `false` as values.
+
+### Required Changes
```

Comment:
Clarify how users can be made aware when their configuration does not apply:
* App logs?
* App stats?

Reply:
I'm not into the details of this topic, maybe someone has ideas on how we could implement this?

Reply:
I guess CF API responds with 4xx and an error message. The error msg can only be as good as the validation rules implemented in CC.

Reply:
Do you have a rough idea about the max size of the map (lets say serialized as json file)? 0.1k, 1k, 10k, 100k, ...?

Reply:
That's an excellent question, I was hoping you can tell me that 😄 

The places where the size matters is probably in the databases which have to store it and the nats message (since they sent frequently). Since there are already [fields](https://github.com/cloudfoundry/cloud_controller_ng/blob/e719d017ea4625397a97c2c14352ebdee66febe9/app/models/runtime/droplet_model.rb#L47-L48) similar to this one: what is the limit on those?

Reply:
> I guess CF API responds with 4xx and an error message.

Yes, when the user submits a new route / updates an existing one the CC would do some validation. But it could still be the case that invalid data makes it to gorouter, how do we handle that?

Reply:
> invalid data makes it to gorouter, how do we handle that?

Maybe log in gorouter and then just go with the foundation default? I wish we could force an app log here to warn the app dev. But maybe we are over-engineering this.

Reply:
Okay so for now we will only log this on gorouter as an error log which is something the operator will have to investigate. If this becomes too much of an issue we might re-visit the topic and implement a way for app devs to see those errors themselves.

@stephanme regarding the limits: can you provide some insights on what an acceptable size would be for CC?

Reply:
> Do you have a rough idea about the max size of the map (lets say serialized as json file)? 0.1k, 1k, 10k, 100k, ...?

I think 10K is a reasonable size that should fit all current needs.